### DISCUSSION TITLE
Image inside a row with the right aspect ratio

### DISCUSSION THREAD

**Discussion by PoshoDev** [zauberzeug/nicegui#3603] [Q&A] [answered]
### Question

Hello, 👋

I'm trying to put an `ui.image` inside of a `ui.row` layout that has a fixed height _while keeping the image's aspect ratio accurate_. I have no idea how this task became a 2+ hour ordeal, so I think it's time to ask for help, if you don't mind. 😅

The main takeaway is that due to the use-case of my application, in our case, the **width of the image should not be hard-coded** since we don't know the dimensions of the image but the aspect ratio should remain.

Here is an MRC of my layout:

```python
from nicegui import ui

WIDTH = 300
HEIGHT = 600

ROW_HEIGHT = 64
MY_SVG = f"https://placeholder.pics/svg/{WIDTH}x{HEIGHT}"

def my_layout() -> ui.row:
    with ui.row().style(f"max-height: {ROW_HEIGHT}px") \
    .classes("w-full border flex-nowrap") \
    .style("border-color: red; align-items: center") as row:
        ui.button(icon="menu")
        ui.space()
        with ui.row().style("padding: 0px;"):
            ui.button(icon="search")
    return row
```

Here are a few of the attempts I've tried:

```python
# Attempt 1
with my_layout():
    ui.image(MY_SVG)  # Just an image.
```
![image](https://github.com/user-attachments/assets/67c5292a-af96-4d45-ae26-426ab586df72)

```python
# Attempt 2
with my_layout():
    ui.image(MY_SVG).style("max-width: 100%; max-height: 100%;")  # Try to limit the size of the image.
```
![image](https://github.com/user-attachments/assets/b7a80ef7-dea0-4d96-86ed-edfa6c9ef29a)

```python
# Attempt 3
with my_layout():
    ui.image(MY_SVG).style(f"height: {ROW_HEIGHT}px").props("fit='contain'") \
    .style("border-color: blue").classes("border")  # Match the height of the parent row using 'contain'.
```
![image](https://github.com/user-attachments/assets/8b05ce09-1d13-4c77-b7fb-a09f87730d5a)

```python
# Attempt 4
with my_layout():
    ui.image(MY_SVG).style(f"height: {ROW_HEIGHT}px").props("fit='shrink-down'")  # Match the height of the parent row using 'shrink-down'.
```
![image](https://github.com/user-attachments/assets/28f31ad7-b1c4-4965-b930-de34099be661)

```python
# Attempt 5
with my_layout():
    ui.image(MY_SVG).style(f"height: {ROW_HEIGHT}px; width: {ROW_HEIGHT}px")
```
![image](https://github.com/user-attachments/assets/4445a469-5c21-4213-ae4a-f10bd52ef523)

```python
# Attempt 6
with my_layout():
    ui.image(MY_SVG).style(f"height: {ROW_HEIGHT}px; width: {ROW_HEIGHT}px").props("fit='contain'") \
    .style("border-color: blue").classes("border")  # Match the height of the parent row using 'contain'.
```
![image](https://github.com/user-attachments/assets/fe449b0c-bbea-4bac-a666-8613c8cd63b7)

```python
# Attempt 7
with my_layout():
    ui.image(MY_SVG).style(f"height: {ROW_HEIGHT}px; width: {ROW_HEIGHT}px").props("fit='shrink-down'")  # Match the height of the parent row using 'shrink-down'.
```
![image](https://github.com/user-attachments/assets/f9d19163-1dbf-4f74-b66f-7c1e854f46c8)

```python
# Attempt 8
with my_layout():
    ui.image(MY_SVG).style(f"max-height: {ROW_HEIGHT}px; max-width: {ROW_HEIGHT}px").props("fit='contain'") \
    .style("border-color: blue").classes("border")  # Match the height of the parent row using 'contain'.
```
![image](https://github.com/user-attachments/assets/ef0bf242-121f-40c5-ac08-c8d8c5308410)

```python
# Attempt 9
with my_layout():
    ui.image(MY_SVG).style(f"height: {ROW_HEIGHT}px; width: auto").props("fit='contain'")  # Match the height of the parent row using 'shrink-down'.
```
![image](https://github.com/user-attachments/assets/16083769-d177-4c11-a2eb-17e5c0f2376d)

Needless to say I've scoured through Quasar documentation, but I can't figure for the life of me how to do this right. Any guidance would be appreciated. 😁

**Comment by falkoschindler** [comment, marked as answer]
Hi @PoshoDev,

I see two possibilities:

- Use [QImg's](https://quasar.dev/vue-components/img) "height" prop instead to set the image height, "fit=scale-down" and "position" to place it inside the occupied space:

    ```py
    with ui.row(wrap=False).classes('w-full h-32 items-center border-2 border-blue-500'):
        ui.button('Button')
        ui.image('https://dummyimage.com/640x480').props(r'height=100% fit=scale-down position="0 0"')
    ```

- Alternatively you can use `ui.interactive_image`, which behaves a bit more ordinarily when it comes to layout and sizing:
    ```py
    with ui.row(wrap=False).classes('w-full h-32 items-center border-2 border-blue-500'):
        ui.button('Button')
        ui.interactive_image('https://dummyimage.com/640x480').classes('h-full')
    ```

In both cases you need to set a height to the `ui.row` container. A maximum height won't be enough for its children to inherit a height.

**Reply by PoshoDev** [reply]
Hello @falkoschindler,

Nice seeing you again! 🙂

**TL;DR:** It worked, thank you!

Your first example stills brings up my issue with the unnecessary space.
```python
with ui.row(wrap=False).classes('w-full h-32 items-center border-2 border-blue-500'):
    ui.button('Button')
    ui.space()  # Added an empty space to push the image to the right.
    ui.image('https://dummyimage.com/640x480').props(r'height=100% fit=scale-down position="0 0"') \
    .classes('border border-4 border-green-500')  # Added a border to the image.
```
![image](https://github.com/user-attachments/assets/4ba10c12-dbd6-4837-a2d3-27118d230690)

However, your second example making use of the `ui.interactive_image` solves it very well!
```python
with ui.row(wrap=False).classes('w-full h-32 items-center border-2 border-blue-500'):
    ui.button('Button')
    ui.space()  # Added an empty space to push the image to the right.
    ui.interactive_image('https://dummyimage.com/640x480').classes('h-full') \
    .classes('border border-4 border-green-500')  # Added a border to the image.
```
![image](https://github.com/user-attachments/assets/5f82539e-63fc-4771-b394-5fef5de2ab84)

I can now use my original layout (although with a height `class` instead of `style`) with the desired behavior:

```python
from nicegui import ui

WIDTH = 300
HEIGHT = 600

ROW_HEIGHT = 64
MY_SVG = f"https://placeholder.pics/svg/{WIDTH}x{HEIGHT}"

def my_layout() -> ui.row:
    with ui.row(wrap=False).classes(f"w-full border h-{ROW_HEIGHT}") \
    .style("border-color: red; align-items: center") as row:
        ui.button(icon="menu")
        ui.space()
        with ui.row().style("padding: 0px;"):
            ui.button(icon="search")
    return row
    
with my_layout():
    ui.interactive_image(MY_SVG).classes('h-full') \
    .classes('border border-4 border-green-500')  # Added a border to the image.
    
if __name__ in {'__main__', '__mp_main__'}:
    ui.run()
```
![image](https://github.com/user-attachments/assets/c08468d1-640e-44ad-bbe2-4162c46cc04c)

I would've never guessed that the solution would've been on an `ui.interactive_image` on my own.

Thank you so much once again! 😁

**Reply by python-and-novella** [reply]
Thanks @falkoschindler 
Second is better:
```python3
with ui.row(wrap=False).classes('w-full h-32 items-center border-2 border-blue-500'):
    ui.button('Button')
    ui.space()
    ui.button('Button')
    ui.interactive_image('https://dummyimage.com/640x480').classes('h-full')
```
![image](https://github.com/user-attachments/assets/fcdcc3f2-1c69-4e15-aa8c-f547e3d3db05)
